File: AlgorithmsProject/src/in/algorithms/bst/BSTOperations.scala
package in.algorithms.bst

class BSTOperations {

  def inOrder(root: BSTNode): Unit = {
    if (root != null) {
      inOrder(root.lchild)
      System.out.print(root.value + " ")
      inOrder(root.rchild)
    }
  }

  def preOrder(root: BSTNode): Unit = {
    if (root != null) {
      System.out.print(root.value + " ")
      preOrder(root.lchild)
      preOrder(root.rchild)

    }
  }

  def postOrder(root: BSTNode): Unit = {
    if (root != null) {
      postOrder(root.lchild)
      postOrder(root.rchild)
      System.out.print(root.value + " ")

    }
  }

  def checkIfBST(root: BSTNode): Boolean = {
    var isBST = true
    var maxCurrent: Int = Integer.MIN_VALUE
    def BSTChecker(root: BSTNode): Unit = {
      if (root != null && isBST) {
        BSTChecker(root.lchild)
        if (root.value < maxCurrent) {
          isBST = false
        } else {
          maxCurrent = root.value
        }
        BSTChecker(root.rchild)
      }
    }
    BSTChecker(root)
    isBST
  }

  def convertToBST(root: BSTNode): BSTNode = {
    var head: BSTNode = null
    var prev: BSTNode = null
    def BSTConvert(root: BSTNode): Unit = {
      if (root == null)
        return
      else {
        BSTConvert(root.lchild)
        if (prev == null)
          head = root
        else {
          root.lchild = prev
          prev.rchild = root
        }
        prev = root
        BSTConvert(root.rchild)
      }
    }
    BSTConvert(root)
    head
  }

  def printInSpiralModel(root: BSTNode, initialLeftToRightValue: Boolean): Unit = {

    var leftToRight = initialLeftToRightValue
    val height = getTreeHeight(root)
    println("Height is " + height)

    def printSpiral(root: BSTNode, height: Int): Unit = {
      if (root == null)
        return
      else if (height == 1)
        print(root.value + " ")
      else {
        if (leftToRight) {
          printSpiral(root.lchild, height - 1)
          printSpiral(root.rchild, height - 1)
        } else {
          printSpiral(root.rchild, height - 1)
          printSpiral(root.lchild, height - 1)
        }
      }
    }

    for (i <- 1 to height) {
      printSpiral(root, i)
      leftToRight = !leftToRight
    }

  }

  def getTreeHeight(root: BSTNode): Int = {
    if (root == null)
      0
    else {
      1 + Math.max(getTreeHeight(root.lchild), getTreeHeight(root.rchild))
    }
  }

  def giveMirrorTree(root: BSTNode): BSTNode = {
    if (root == null)
      return null
    else {
      val left = giveMirrorTree(root.lchild);
      val right = giveMirrorTree(root.rchild);
      return new BSTNode(root.value, right, left)
    }
  }

  def convertToMirror(root: BSTNode): Unit = {
    if (root == null) return
    else {
      convertToMirror(root.lchild)
      convertToMirror(root.rchild)
      var temp = root.lchild
      root.lchild = root.rchild
      root.rchild = temp

    }

  }

  def maxFromLeafToRoot(root: BSTNode): Int = {

    if (root == null) 0
    else
      root.value + Math.max(maxFromLeafToRoot(root.lchild), maxFromLeafToRoot(root.rchild))
  }

  def maxFromLeafToLeaf(root: BSTNode): Int = {
    var highestSum = 0
    def max(root: BSTNode): Int = {
      if (root == null) 0
      else {
        val leftMax = max(root.lchild)
        val rightMax = max(root.rchild)
        if (highestSum < root.value + leftMax + rightMax)
          highestSum = root.value + leftMax + rightMax
        root.value + Math.max(max(root.lchild), max(root.rchild))
      }
    }
    max(root)
    highestSum
  }
  
  def printHeightWise(root : BSTNode, depth : Int) : Unit = {
    if(root==null) return
    else {
      if(depth == 1) print(root.value + " ")
      else {
        printHeightWise(root.lchild, depth-1)
        printHeightWise(root.rchild, depth-1)
      }
    }
 
    
  }
}

File: AlgorithmsProject/src/in/algorithms/bst/BSTOperations.java
package in.algorithms.bst;

import java.util.ArrayList;
import java.util.List;

public class BSTOperations {

    public static <T extends Comparable<T>> BSTNode<T> insert(BSTNode<T> root, T value) {
        if (root == null) {
            return new BSTNode<>(value);
        }
        int cmp = value.compareTo(root.value);
        if (cmp < 0) {
            root.left = insert(root.left, value);
        } else if (cmp > 0) {
            root.right = insert(root.right, value);
        }
        return root;
    }

    public static <T extends Comparable<T>> boolean search(BSTNode<T> root, T value) {
        if (root == null) return false;
        int cmp = value.compareTo(root.value);
        if (cmp == 0) return true;
        return cmp < 0 ? search(root.left, value) : search(root.right, value);
    }

    public static <T extends Comparable<T>> BSTNode<T> findMin(BSTNode<T> root) {
        if (root == null) return null;
        while (root.left != null) {
            root = root.left;
        }
        return root;
    }

    public static <T extends Comparable<T>> BSTNode<T> findMax(BSTNode<T> root) {
        if (root == null) return null;
        while (root.right != null) {
            root = root.right;
        }
        return root;
    }

    public static <T extends Comparable<T>> BSTNode<T> delete(BSTNode<T> root, T value) {
        if (root == null) return null;
        int cmp = value.compareTo(root.value);
        if (cmp < 0) {
            root.left = delete(root.left, value);
        } else if (cmp > 0) {
            root.right = delete(root.right, value);
        } else {
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;
            BSTNode<T> minRight = findMin(root.right);
            root.value = minRight.value;
            root.right = delete(root.right, minRight.value);
        }
        return root;
    }

    public static <T extends Comparable<T>> List<T> inOrder(BSTNode<T> root) {
        List<T> result = new ArrayList<>();
        inOrderHelper(root, result);
        return result;
    }

    private static <T extends Comparable<T>> void inOrderHelper(BSTNode<T> node, List<T> result) {
        if (node != null) {
            inOrderHelper(node.left, result);
            result.add(node.value);
            inOrderHelper(node.right, result);
        }
    }

    public static <T extends Comparable<T>> List<T> preOrder(BSTNode<T> root) {
        List<T> result = new ArrayList<>();
        preOrderHelper(root, result);
        return result;
    }

    private static <T extends Comparable<T>> void preOrderHelper(BSTNode<T> node, List<T> result) {
        if (node != null) {
            result.add(node.value);
            preOrderHelper(node.left, result);
            preOrderHelper(node.right, result);
        }
    }

    public static <T extends Comparable<T>> List<T> postOrder(BSTNode<T> root) {
        List<T> result = new ArrayList<>();
        postOrderHelper(root, result);
        return result;
    }

    private static <T extends Comparable<T>> void postOrderHelper(BSTNode<T> node, List<T> result) {
        if (node != null) {
            postOrderHelper(node.left, result);
            postOrderHelper(node.right, result);
            result.add(node.value);
        }
    }

    public static <T extends Comparable<T>> int size(BSTNode<T> root) {
        if (root == null) return 0;
        return 1 + size(root.left) + size(root.right);
    }

    public static <T extends Comparable<T>> int height(BSTNode<T> root) {
        if (root == null) return 0;
        return 1 + Math.max(height(root.left), height(root.right));
    }
}
